synthesizedBaseline uses incorrect line under edge for vertical-lr and sideways items

sammygill opened the pull request:
#### bc94fc496ea98de389ed201f55ab2b4f089a0f7e
<pre>
synthesizedBaseline uses incorrect line under edge for vertical-lr and sideways items
<a href="https://bugs.webkit.org/show_bug.cgi?id=263461">https://bugs.webkit.org/show_bug.cgi?id=263461</a>
rdar://117273063

Reviewed by Alan Baradlay.

To synthesize a baseline, the spec says the following: &quot;synthesize the
alphabetic baseline from the line-under line, and the central baseline
by averaging the positions of the two edges or lines.&quot;

When the writing mode is vertical-lr and text-orientation is sideways,
we must synthesize the alphabetic baseline. This means that the baseline
for the item in this writing-mode is its block-start and not the
block-end like in vertical-rl writing mode.

We can first determine the baseline that we are synthesizing according
to the logic in: <a href="https://drafts.csswg.org/css-inline-3/#dominant-baseline-property">https://drafts.csswg.org/css-inline-3/#dominant-baseline-property</a>
If the baseline type is alphabetic and we are in a vertical-lr writing
mode according to the formatting context root, then we can return 0_lu
for the synthesized baseline.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/baseline-synthesis-vert-lr-line-under-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/baseline-synthesis-vert-lr-line-under.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-container-baseline-001-expected.txt:
* Source/WebCore/rendering/RenderBox.cpp:
 (WebCore::synthesizedBaseline):

Canonical link: <a href="https://commits.webkit.org/269668@main">https://commits.webkit.org/269668@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d131b8a3f9733277e8a9686699e3b0fe808850b0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23195 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1282 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24316 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/25120 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21437 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/23463 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/2790 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23734 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/22289 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23435 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/882 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20103 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25958 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/704 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20988 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/27156 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/21154 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21252 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/25031 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/693 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18459 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/627 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/20765 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5540 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1104 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/902 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -5680,14 +5680,33 @@ LayoutBoxExtent RenderBox::scrollPaddingForViewportRect(const LayoutRect& viewpo
+        auto isHorizontalWritingMode = parentStyle.isHorizontalWritingMode();
+        auto textOrientation = parentStyle.textOrientation();
+
+        if (isHorizontalWritingMode || (!isHorizontalWritingMode && textOrientation == TextOrientation::Sideways))
```
> if (isHorizontalWritingMode || (!isHorizontalWritingMode && textOrientation == TextOrientation::Sideways))

^^This looks slightly redundant.
`if (isHorizontalWritingMode || textOrientation == TextOrientation::Sideways)`

sammygill replied:
Fixed

alanbaradlay commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -5680,14 +5680,33 @@ LayoutBoxExtent RenderBox::scrollPaddingForViewportRect(const LayoutRect& viewpo
+
+        if (isHorizontalWritingMode || (!isHorizontalWritingMode && textOrientation == TextOrientation::Sideways))
+            return FontBaseline::AlphabeticBaseline;
+        if (!isHorizontalWritingMode && (textOrientation == TextOrientation::Upright || textOrientation == TextOrientation::Mixed))
```
this too as horizontal mode is already handled.
`if (textOrientation == TextOrientation::Upright || textOrientation == TextOrientation::Mixed)`

alanbaradlay replied:
or you could turn this into something like this

```
if (isHorizontalWritingMode)
  return FontBaseline::AlphabeticBaseline;
else {
  switch (parentStyle.textOrientation() {
    case TextOrientation::Sideways:
      return FontBaseline::AlphabeticBaseline;
   ...
}
```
to avoid consulting parentStyle.textOrientation() in the common case of horizontal writing mode, but since this is certainly not a hot codepath, your original approach is okay too.

sammygill replied:
Fixed

sammygill committed:
synthesizedBaseline uses incorrect line under edge for vertical-lr and sideways items
https://bugs.webkit.org/show_bug.cgi?id=263461
rdar://117273063

Reviewed by Alan Baradlay.

To synthesize a baseline, the spec says the following: "synthesize the
alphabetic baseline from the line-under line, and the central baseline
by averaging the positions of the two edges or lines."

When the writing mode is vertical-lr and text-orientation is sideways,
we must synthesize the alphabetic baseline. This means that the baseline
for the item in this writing-mode is its block-start and not the
block-end like in vertical-rl writing mode.

We can first determine the baseline that we are synthesizing according
to the logic in: https://drafts.csswg.org/css-inline-3/#dominant-baseline-property
If the baseline type is alphabetic and we are in a vertical-lr writing
mode according to the formatting context root, then we can return 0_lu
for the synthesized baseline.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/baseline-synthesis-vert-lr-line-under-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/baseline-synthesis-vert-lr-line-under.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-container-baseline-001-expected.txt:
* Source/WebCore/rendering/RenderBox.cpp:
 (WebCore::synthesizedBaseline):

Canonical link: https://commits.webkit.org/269668@main
